jsf-api/src/main/java/javax/faces/component/search/SearchExpressionHandler.java

```diff
@@ -0,0 +1,277 @@
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package javax.faces.component.search;
+
+import java.util.List;
+import javax.faces.component.ContextCallback;
+import javax.faces.component.UIComponent;
+import javax.faces.context.FacesContext;
+
+/**
+ * <p class="changed_added_2_3">The <strong>SearchExpressionHandler</strong> is responsible for
+ * resolving <em>search expression(s)</em>.
+ * 
+ * <p>A <em>search expression</em> consists of either an identifier
+ * (which is matched exactly against the <code>id</code> property of
+ * a {@link UIComponent}, or an keyword (like <code>@this</code> or <code>@form</code>),
+ * or a series of such identifiers and keywords linked by
+ * the {@link UINamingContainer#getSeparatorChar} character value.
+ * The search algorithm should operates as follows, though alternate
+ * alogrithms may be used as long as the end result is the same:</p>
+ * 
+ * <ul>
+ *   <li>
+ *     Identify the {@link UIComponent} that will be the base for searching:
+ *     <ul>
+ *       <li>
+ *           If the search expression begins with the separator character
+ *           (called an "absolute" search expression),
+ *           the base will be the root {@link UIComponent} of the component
+ *           tree. The leading separator character will be stripped off,
+ *           and the remainder of the search expression will be treated as
+ *           a "relative" search expression as described below.
+ *       </li>
+ *       <li>
+ *           Otherwise, the {@link SearchExpressionContext#getSource()} will be used.
+ *       </li>
+ *     </ul>
+ *   </li>
+ *   <li>
+ *     The search expression (possibly modified in the previous step) is now
+ *     a "relative" search expression that will be used to locate the
+ *     component (if any) based on the identifier and/or keywords:
+ *     <ul>
+ *       <li>
+ *          The expression will be splitted by {@link UINamingContainer#getSeparatorChar} into "commands".
+ *          The commands will be resolved one by one.
+ *          For the first command, the source component, like mentioned above, will be used to start the lookup.
+ *          For all further commands, the previous resolved component, from the previous command, will be used the start the lookup.
+ *       </li>
+ *       <li>
+ *           If the command starts with the {@link #KEYWORD_PREFIX}, the {@link SearchKeywordResolver}s
+ *           will used the resolve the command.
+ *       </li>
+ *       <li>
+ *           Otherwise, If the command starts not with the {@link #KEYWORD_PREFIX}, the component will be resolved
+ *           based on the component id. Similar to {@link UIComponent#findComponent}.

```

Comment:
Poor grammar.

> ```
> Otherwise, if the command does not start with {@link #KEYWORD_PREFIX}, then the component will be resolved
> based on the component ID using {@link UIComponent#findComponent}.
> ```